inc/api/namespace.php

```diff
@@ -231,27 +234,24 @@ function get_graph_data( $start, $end, $resolution = '1 day', ?Filter $filter =
 			uniqCombined64If(endpoint_id, endpoint_metrics['pageViews'] < 2) as bounce
 		FROM analytics
 		WHERE
-			attributes['blogId'] = %s
-			AND event_timestamp >= toDateTime64(intDiv(%d,1000),3)
-			AND event_timestamp <= toDateTime64(intDiv(%d,1000),3)
+			blog_id = {blog_id:String}
+			AND event_timestamp >= toDateTime64({start:UInt64},3)
+			AND event_timestamp <= toDateTime64({end:UInt64},3)
 			AND event_type = 'pageView'
 			{$query_where}
 		GROUP BY
 			`date`
 			WITH ROLLUP
-		ORDER BY `date` ASC",
-		get_current_blog_id(),
-		(int) sprintf( '%d000', $start ),
-		(int) sprintf( '%d999', $end )
-	);
+		ORDER BY `date` ASC
+			WITH FILL FROM toDateTime({start:UInt64}) STEP INTERVAL {$resolution}";
 
-	$key = sprintf( 'analytics:stats:%s', sha1( serialize( $query ) ) );
+	$key = sprintf( 'analytics:stats:%s', sha1( serialize( [ $query, $query_params ] ) ) );
```

Comment:
This may be worth abstracting; `function key_for_query( string $query, ?array $params ) : string;`

Reply:
Yeah there's a bunch I could abstract really but this pattern is only used in maybe 2 places. There's a bunch of other improvements in general to make anyway, like splitting some functions / queries up into multiple API endpoints etc..